File: backend/app/schemas/history.py
# backend/app/schemas/history.py
"""
Pydantic schemas for history API endpoints.
These schemas define the structure of history items returned by the API.
"""

from typing import List, Optional
from uuid import UUID
from datetime import datetime
from pydantic import BaseModel, Field, ConfigDict
from enum import Enum


class HistoryItemType(str, Enum):
    """Type of history item."""
    SUMMARY = "summary"
    QUIZ = "quiz"


class SummaryHistoryItem(BaseModel):
    """Schema for a summary item in history list."""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    document_id: UUID
    document_title: str = Field(..., description="Title/filename of the source document")
    summary_preview: Optional[str] = Field(None, description="First 200 characters of summary")
    generated_at: datetime
    ai_model: str
    type: HistoryItemType = HistoryItemType.SUMMARY


class QuizHistoryItem(BaseModel):
    """Schema for a quiz item in history list."""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    document_id: UUID
    document_title: str = Field(..., description="Title/filename of the source document")
    title: str = Field(..., description="Quiz title")
    status: str
    total_questions: int
    created_at: datetime
    ai_model: str
    type: HistoryItemType = HistoryItemType.QUIZ


class SummaryHistoryResponse(BaseModel):
    """Response wrapper for summary history list."""
    data: List[SummaryHistoryItem]
    message: str = "Summaries retrieved successfully"
    status: str = "success"
    total: int = Field(..., description="Total number of summaries")


class QuizHistoryResponse(BaseModel):
    """Response wrapper for quiz history list."""
    data: List[QuizHistoryItem]
    message: str = "Quizzes retrieved successfully"
    status: str = "success"
    total: int = Field(..., description="Total number of quizzes")


class CombinedHistoryItem(BaseModel):
    """Schema for a combined history item (either summary or quiz)."""
    model_config = ConfigDict(from_attributes=True)
    
    id: UUID
    document_id: UUID
    document_title: str
    title: Optional[str] = None  # For quizzes
    preview: Optional[str] = None  # For summaries (first 200 chars)
    type: HistoryItemType
    status: Optional[str] = None  # For quizzes
    total_questions: Optional[int] = None  # For quizzes
    created_at: datetime
    ai_model: str


class CombinedHistoryResponse(BaseModel):
    """Response wrapper for combined history list."""
    data: List[CombinedHistoryItem]
    message: str = "History retrieved successfully"
    status: str = "success"
    total: int = Field(..., description="Total number of history items")

File: frontend/src/services/history.ts
// frontend/src/services/history.ts

import axios from 'axios'

export type HistoryItemType = 'summary' | 'quiz'

export interface SummaryHistoryItem {
  id: string
  document_id: string
  document_title: string
  summary_preview: string | null
  generated_at: string
  ai_model: string
  type: 'summary'
}

export interface QuizHistoryItem {
  id: string
  document_id: string
  document_title: string
  title: string
  status: string
  total_questions: number
  created_at: string
  ai_model: string
  type: 'quiz'
}

export interface CombinedHistoryItem {
  id: string
  document_id: string
  document_title: string
  title: string | null
  preview: string | null
  type: HistoryItemType
  status: string | null
  total_questions: number | null
  created_at: string
  ai_model: string
}

export interface SummaryHistoryResponse {
  data: SummaryHistoryItem[]
  message: string
  status: string
  total: number
}

export interface QuizHistoryResponse {
  data: QuizHistoryItem[]
  message: string
  status: string
  total: number
}

export interface CombinedHistoryResponse {
  data: CombinedHistoryItem[]
  message: string
  status: string
  total: number
}

/**
 * Get summary history for authenticated user
 */
export const getSummaryHistory = async (
  accessToken: string,
  limit: number = 50,
  offset: number = 0
): Promise<SummaryHistoryResponse> => {
  const headers: Record<string, string> = {
    Authorization: `Bearer ${accessToken}`,
  }

  try {
    const response = await axios.get<SummaryHistoryResponse>(
      `/api/v1/history/summaries?limit=${limit}&offset=${offset}`,
      { headers }
    )
    return response.data
  } catch (error: any) {
    if (axios.isAxiosError(error)) {
      const errorMessage = error.response?.data?.detail || error.message
      throw new Error(`Failed to fetch summary history: ${errorMessage}`)
    }
    throw new Error(`An unexpected error occurred: ${error.message}`)
  }
}

/**
 * Get quiz history for authenticated user
 */
export const getQuizHistory = async (
  accessToken: string,
  limit: number = 50,
  offset: number = 0
): Promise<QuizHistoryResponse> => {
  const headers: Record<string, string> = {
    Authorization: `Bearer ${accessToken}`,
  }

  try {
    const response = await axios.get<QuizHistoryResponse>(
      `/api/v1/history/quizzes?limit=${limit}&offset=${offset}`,
      { headers }
    )
    return response.data
  } catch (error: any) {
    if (axios.isAxiosError(error)) {
      const errorMessage = error.response?.data?.detail || error.message
      throw new Error(`Failed to fetch quiz history: ${errorMessage}`)
    }
    throw new Error(`An unexpected error occurred: ${error.message}`)
  }
}

/**
 * Get combined history (summaries + quizzes) for authenticated user
 */
export const getCombinedHistory = async (
  accessToken: string,
  limit: number = 50,
  offset: number = 0
): Promise<CombinedHistoryResponse> => {
  const headers: Record<string, string> = {
    Authorization: `Bearer ${accessToken}`,
  }

  try {
    const response = await axios.get<CombinedHistoryResponse>(
      `/api/v1/history?limit=${limit}&offset=${offset}`,
      { headers }
    )
    return response.data
  } catch (error: any) {
    if (axios.isAxiosError(error)) {
      const errorMessage = error.response?.data?.detail || error.message
      throw new Error(`Failed to fetch history: ${errorMessage}`)
    }
    throw new Error(`An unexpected error occurred: ${error.message}`)
  }
}

// ============================================================================
// Quiz Review Types and Functions (Story 5.2)
// ============================================================================

export interface QuestionReviewItem {
  id: string
  question_text: string
  question_type: string
  options: string[] | null
  correct_answer: string
  explanation: string | null
  user_answer: string | null
  is_correct: boolean | null
  order_index: number
}

export interface QuizReviewDetail {
  id: string
  document_id: string
  document_title: string
  title: string
  status: string
  total_questions: number
  score: number
  score_percentage: number
  ai_model: string
  created_at: string
  questions: QuestionReviewItem[]
}

export interface QuizReviewResponse {
  data: QuizReviewDetail
  message: string
  status: string
}

/**
 * Get detailed quiz review with user's answers
 */
export const getQuizReview = async (
  accessToken: string,
  quizId: string
): Promise<QuizReviewResponse> => {
  const headers: Record<string, string> = {
    Authorization: `Bearer ${accessToken}`,
  }

  try {
    const response = await axios.get<QuizReviewResponse>(
      `/api/v1/history/quizzes/${quizId}`,
      { headers }
    )
    return response.data
  } catch (error: any) {
    if (axios.isAxiosError(error)) {
      const errorMessage = error.response?.data?.detail || error.message
      throw new Error(`Failed to fetch quiz review: ${errorMessage}`)
    }
    throw new Error(`An unexpected error occurred: ${error.message}`)
  }
}
